### PR TITLE
Refactor insertSigToBuilder and handleSig, avoid unused allocations

### PR DESCRIPTION
**Pull request by Aharonee**
- Change insertSigToBuilder from Worker method to builder method
- Reuse insertSig in handleSig as well, with the newly defined custom errors
- Implement getMessage to fetch stateproof message from DB instead of using getBuilder

### CODE REVIEW

File: stateproof/builder.go [hunk 1]
```diff
@@ -252,32 +271,34 @@ func (spw *Worker) getAllOnlineBuilderRounds() ([]basics.Round, error) {
 	return rnds, err
 }
 
-func (spw *Worker) insertSigToBuilder(builderForRound *builder, sig *pendingSig) {
-	rnd := builderForRound.Round
-	pos, ok := builderForRound.AddrToPos[sig.signer]
+var errAddressNotInCommittee = errors.New("cannot find address in builder")              // Address was not a part of Top N accounts in this StateProof's committee
```
Comment (algoidan) [inline stateproof/builder.go:274]: nits: can you please avoid using the word committee? avoid confusion with algorand's consensus

File: stateproof/builder.go [hunk 2]
```diff
@@ -252,32 +271,34 @@ func (spw *Worker) getAllOnlineBuilderRounds() ([]basics.Round, error) {
...
+var errSigAlreadyPresentAtPos = errors.New("signature already present at this position") // Signature already present at this position
+var errSignatureVerification = errors.New("error while verifying signature")             // Signature failed cryptographic verification
+
+func (b *builder) insertSig(s *pendingSig, verify bool) error {
+	rnd := b.Round
+	pos, ok := b.AddrToPos[s.signer]
 	if !ok {
-		spw.log.Warnf("insertSigToBuilder: cannot find %v in round %d", sig.signer, rnd)
-		return
+		return fmt.Errorf("insertSigToBuilder: %w (%v not in participants for round %d)", errAddressNotInCommittee, s.signer, rnd)
```
Comment (algoidan) [inline stateproof/builder.go:283]: nits: change the function name

**Review by algoidan — APPROVED** [review]
LGTM.
Have a few minor nits comments.

File: stateproof/builder.go [hunk 3]
```diff
@@ -367,36 +388,30 @@ func (spw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.Forw
...
 		return network.Ignore, nil
 	}
-
-	if err := builderForRound.IsValid(pos, &sfa.Sig, true); err != nil {
+	if errors.Is(err, errFailedToAddSigAtPos) {
+		return network.Ignore, err
+	}
+	if errors.Is(err, errAddressNotInVoters) || errors.Is(err, errSignatureVerification) {
 		return network.Disconnect, err
 	}
```
Comment (almog-t) [inline stateproof/builder.go:407]: Should probably positively verify that err is nil here, to avoid future bugs involving modification of insertSig.

**Review by almog-t — APPROVED** [review]

**Pull request merged**